### PR TITLE
feat: implement partition chunk rollover + ids and timestamps

### PR DESCRIPTION
**Pull request by alamb**
Closes https://github.com/influxdata/influxdb_iox/issues/598 and closes https://github.com/influxdata/influxdb_iox/issues/579

This PR adds adds the `rollover_chunk` and `drop_chunk` apis to `Partition` as well as adding Chunk Id's and timestamps. 

Most of the PR's size is tests.

Notes:
1. There are also likely corner cases related to query (aka needing to combine data from multiple chunks for a partition before grouping or ordering, for example) that will need thought -- I filed https://github.com/influxdata/influxdb_iox/issues/602 to track this work. 

2. This doesn't include any end-to-end tests for rollover controlled by the API, as I believe @pauldix is planning on re-working that API shortly.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: mutable_buffer/src/chunk.rs [hunk 1]
```diff
@@ -60,9 +56,26 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
 
 #[derive(Debug)]
 pub struct Chunk {
-    /// Chunk key for all rows in this the chunk
+    /// The id for this chunk
```
Comment (alamb) [inline mutable_buffer/src/chunk.rs:59]: Here are the new fields for each Chunk

File: mutable_buffer/src/database.rs [hunk 2]
```diff
@@ -1103,15 +1103,6 @@ mod tests {
         v.iter().map(|s| s.to_string()).collect::<BTreeSet<_>>()
     }
 
-    // Abstract a bit of boilerplate around table assertions and improve the failure
-    // output. The default failure message uses Debug formatting, which prints
-    // newlines as `\n`. This prints the pretty_format_batches using Display so
-    // it's easier to read the tables.
-    fn assert_table_eq(table: &str, chunks: &[RecordBatch]) {
```
Comment (alamb) [inline mutable_buffer/src/database.rs:1110]: I factored this test helper out into `arrow_deps`

File: mutable_buffer/src/partition.rs [hunk 3]
```diff
@@ -60,56 +93,589 @@ impl Partition {
...
+                    Some(&partition.mutable_chunk)
+                } else {
+                    None
+                }
+            })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (alamb) [inline mutable_buffer/src/partition.rs:235]: here are all the tests showing how this code works

File: mutable_buffer/src/chunk.rs [hunk 4]
```diff
@@ -60,9 +56,26 @@ pub type Result<T, E = Error> = std::result::Result<T, E>;
...
+    pub id: u64,
+
+    /// partition key for all rows in this chunk
     pub key: String,
 
+    /// Time at which the first data was written into this chunk. Note
+    /// this is not the same as the timestamps on the data itself
+    pub time_of_first_write: Option<Instant>,
```
Comment (pauldix) [inline mutable_buffer/src/chunk.rs:67]: This ins't for timing and more for the actual time. I'd opt for `DateTime<Utc>` in the `chrono` crate. We're using that elsewhere. I don't think we need the monotonically increasing guarantee that `Instant` provides? Same for the next two times.

File: mutable_buffer/src/partition.rs [hunk 5]
```diff
@@ -31,17 +54,27 @@ pub struct Partition {
...
+    /// Use BTreeMap here as it is ordered by id and we need to use
+    /// `iter()` to iterate over chunks in their creation order
+    immutable_chunks: BTreeMap<u64, Arc<Chunk>>,
+
+    id_generator: ChunkIdGenerator,
 }
 
 impl Partition {
     pub fn new(key: impl Into<String>) -> Self {
+        // TODO: for existing partitions, does this need to pick up at preexisting ID?
```
Comment (pauldix) [inline mutable_buffer/src/partition.rs:68]: Yes, a partition can be read from object storage, which will read chunks. So a new chunk created should have the next ID.

File: mutable_buffer/src/partition.rs [hunk 6]
```diff
@@ -60,56 +93,589 @@ impl Partition {
...
+/// information on chunks for this partition
+#[derive(Debug, Default, PartialEq)]
+pub struct PartitionChunkInfo {
+    pub num_immutable_chunks: usize,
+}
+
+/// Responsible for assigning ids to chunks. Eventually, this might
+/// need to start at a number other than 0.
+#[derive(Debug, Default, PartialEq)]
+struct ChunkIdGenerator {
```
Comment (pauldix) [inline mutable_buffer/src/partition.rs:181]: Any reason why this can't just be an atomic u64 on the partition? This feels like abstraction that adds complexity.

File: mutable_buffer/src/partition.rs [hunk 7]
```diff
@@ -60,56 +93,589 @@ impl Partition {
...
     pub fn rollover_chunk(&mut self) -> Arc<Chunk> {
-        todo!();
+        let chunk_id = self.id_generator.next_id();
+        let mut chunk = Chunk::new(self.key(), chunk_id);
+        std::mem::swap(&mut chunk, &mut self.mutable_chunk);
+        chunk.mark_immutable();
+        let chunk = Arc::new(chunk);
+        if !chunk.is_empty() {
+            let existing_value = self.immutable_chunks.insert(chunk.id(), chunk.clone());
+            assert!(existing_value.is_none());
```
Comment (pauldix) [inline mutable_buffer/src/partition.rs:133]: This is nice defensive programming. I have to add more of this to my tool belt 😄

File: mutable_buffer/src/partition.rs [hunk 8]
```diff
@@ -60,56 +93,589 @@ impl Partition {
...
+    ) -> crate::chunk::Result<()> {
+        for chunk in self.iter() {
+            chunk.table_to_arrow(dst, table_name, columns)?
+        }
+        Ok(())
+    }
+
+    /// Return information about the chunks held in this partition
+    #[allow(dead_code)]
+    pub fn chunk_info(&self) -> PartitionChunkInfo {
```
Comment (pauldix) [inline mutable_buffer/src/partition.rs:112]: I think we'll need the chunk information to be reflected in partition metadata. Like how many chunks there are, and each chunk will have its own metadata. But it looks like this is just for testing purposes. Likely worth revisiting as the paritition/chunk object store persistence gets hammered out more.

**Review by pauldix — COMMENTED** [review]
A few questions and comments...

File: mutable_buffer/src/chunk.rs (hunk 4 above)
Comment (alamb) [reply mutable_buffer/src/chunk.rs:67]: Sounds good -- I'll update to use `DateTime<Utc>` -- I just picked `Instant` as it was part of `std`

File: mutable_buffer/src/partition.rs (hunk 5 above)
Comment (alamb) [reply mutable_buffer/src/partition.rs:68]: That makes sense -- I suppose once the mutable buffer is hooked up to read/write from object store that would be the logical time to hook up these ids.

File: mutable_buffer/src/partition.rs (hunk 6 above)
Comment (alamb) [reply mutable_buffer/src/partition.rs:181]: Well, I was thinking that something would eventually have to initialize the chunk id generator in some non trivial way -- however, as you say, that is speculative. I will remove the struct and use a single field instead

File: mutable_buffer/src/partition.rs (hunk 8 above)
Comment (alamb) [reply mutable_buffer/src/partition.rs:112]: 👍

File: mutable_buffer/src/chunk.rs (hunk 4 above)
Comment (alamb) [reply mutable_buffer/src/chunk.rs:67]: done in 9267de881aa03f1fdebfcc4d16f941497df1de72

File: mutable_buffer/src/partition.rs (hunk 6 above)
Comment (alamb) [reply mutable_buffer/src/partition.rs:181]: 86ff8d0f12ecf4d64088fce82ff1a6379d336ea4

**Review by pauldix — APPROVED** [review]
